rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,385 @@
+# RFC 12217 - 2022-08-25 - Handling Discarded Events
+
+We want to give Vector users the assurance that events are never lost unexpectedly. There are
+currently many circumstances in which events are simply discarded without recourse for users of
+Vector to recover those events. This RFC presents a framework for handling those discarded events in
+a uniform fashion in all components.
+
+## Context
+
+- [Support dead letter queue sources](https://github.com/vectordotdev/vector/issues/1772)
+- [Vector sources acknowledge discarded events](https://github.com/vectordotdev/vector/issues/12217)
+- [Reroute discarded events from throttle transform](https://github.com/vectordotdev/vector/issues/13549)
+- Sink Error Handling RFC (internal document)
+
+## Scope
+
+### In scope
+
+- Handling of discarded events in sources, transforms, and sinks.
+
+### Out of scope
+
+- Handling of discarded data in sources before it is translated into events.
+- Handling of discarded sink request data after it is translated from events.
+
+## Pain
+
+Vector has a number of ways in which components may drop events due to errors that are outside of
+the operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+
+This leads to the potential for lost events and even silently dropped events. Users want to be able
+to capture discarded events at various points in the topology in order to forward them to fail-safe
+destinations for diagnosis.
+
+## Proposal
+
+There are three parts to the proposed solution:
+
+1. Add a new standardized output to all components that may discard events after errors.
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
```

Comment:
What will be the format of the events flowing to the `errors` output? Will we follow the precedent set by the `remap` transform of wrapping the event with some additional metadata about the failure? I think it'd be good to call that out explicitly in here.

Reply:
Maybe this should take advantage of the new metadata namespacing too? That feature will only exist for logs, but we could extend it to metrics and traces. Also will we offer any guarantees about the metadata, if we do include it? I could see users starting to route on it based on the type of failure. I'd suggest we push back on that clearly document that the metadata is opaque and just for human consumption for now.

Reply:
To validate your intuition, I would be tempted to route based on the metadata from an error.  Hopefully a specific example will help:

Kafka sinks can throw errors when the broker sees a message that's too large.  I'm finding it difficult to know the exact message size before sending, so it's hard to build a filter upstream.  I could see handling this situation by routing the error.  I might route to a console log or split the message and try again.  

At a minimum, I could also envision using the error data to know whether dropping the message is okay or if I need to trigger alerts.